STM32 Family: update tests and enable RTC support

mprse opened the pull request:
# Description

This PR provides the following changes:
- Enable RTC support for STM32 family,
- Adapt RTC drivers to the new RTC HAL API standards (free function keeps RTC time and disables only access to the backup domain) - skipped this has been already done by @jeromecoutant in this commit https://github.com/ARMmbed/mbed-os/pull/6217/commits/45cbdc889ff1f80e257f7db1eb1146765f633336 - Thank you!.
- RTC test fix: verify if rtc_isenabled() returns 1 in case when init is done and RTC time is set. Some platforms use RTC for low power timer, so RTC may be already enabled.
- RTC test fix: increase test timeout (verified that on some platforms 30s is not enough).

jeromecoutant commented on targets/targets.json:
```diff
@@ -1192,7 +1192,7 @@
         "extra_labels_add": ["STM32F4", "STM32F429", "STM32F429ZI", "STM32F429xx", "STM32F429xI"],
         "macros_add": ["USB_STM_HAL", "USBHOST_OTHER"],
-        "device_has_add": ["ANALOGOUT", "CAN", "LOWPOWERTIMER", "SERIAL_ASYNCH", "SERIAL_FC", "TRNG", "FLASH"],
+        "device_has_add": ["RTC", "ANALOGOUT", "CAN", "LOWPOWERTIMER", "SERIAL_ASYNCH", "SERIAL_FC", "TRNG", "FLASH"],
```
RTC is already inherited from FAMILY_STM32

mprse replied:
RTC has been disabled on the feature branch for all targets (https://github.com/ARMmbed/mbed-os/commit/bdcb3d239e92508d783fef8ae0f7074141f5b2fb). This PR enables RTC only for NUCLEO-F429ZI.

jeromecoutant replied:
So, please, enable RTC back in FAMILY_STM32 level.
I am going to provide test report for all STM32 targets.

Thx

bcostm commented on targets/TARGET_STM/rtc_api.c:
```diff
@@ -159,13 +159,8 @@ void rtc_free(void)
     HAL_PWR_DisableBkUpAccess();
```
This line can be removed as it is the same below.

jeromecoutant commented on targets/TARGET_STM/rtc_api.c:
```diff
@@ -28,7 +28,7 @@
  */
 
-#if DEVICE_RTC
+#if DEVICE_RTC | DEVICE_LOWPOWERTIMER
```
Not so easy...
If MBED_CONF_TARGET_LOWPOWERTIMER_LPTIM is enabled, RTC can be removed.

mprse commented:
Enabled RTC on FAMILY_STM32 level and removed commit which was conditionally enabling RTC for boards with LOWPOWERTIMER.

mprse commented:
> Hi
> During tests, I found 2 issues:
> 1 - with targets without LSE
> 2 - with F1
> I then push the correction on the master branch as this is not specific to this RTC branch
> 
> See #6217 

Waiting for #6217 to be merged on master. Then RTC feature branch will have to be rebased.
With provided fix RTC support should be enabled for entire STM32 family. This patch will have to be modified so only the RTC is re-enabled and only test fixes are provided (RTC driver fix should not be needed).

cmonr commented:
@mprse Feel free to continue. #6217 is now in master.

mprse commented:
> @mprse Feel free to continue. #6217 is now in master.

Thanks @cmonr ! To continue with this one I need to have destination branch rebased on master, which can be found here (rebased on master 8b6a7aacc5d2b90ba40d89c8eeb680ebee81ea18, all conflicts resolved):
https://github.com/mprse/mbed-os/commits/feature-hal-spec-rtc

cmonr commented:
@mprse Thanks for the heads up. I've updated the feature branch.
There was one conflict with a single target within targets/targets.json that I resolved as best I thought.

mprse commented:
@cmonr Thanks for rebasing on master. Rebased this PR and removed fix for `free` function which is now already on master. Updated PR name and description.

Checked that `tests-mbed_hal-rtc_reset` and `tests-mbed_hal-rtc` tests gives positive results on 3 different NUCLEO boards (NUCLEO_F429ZI, NUCLEO_L476RG, NUCLEO_F070RB).

mprse commented:
@adbridge @cmonr Can we run morph here?

adbridge commented:
@jeromecoutant @bcostm @c1728p9 Are you guys happy that your review comments have been addressed ?

adbridge commented:
@mprse This should really be approved by the reviewers before it gets submitted into the CI system. There may also be a delay getting it in as we have code freeze for 5.8 tomorrow so that is the priority.

bcostm commented:
OK for me! and FYI @jeromecoutant is out of office till March 12.

cmonr commented:
@ARMmbed/team-st-mcd Do y'all want to wait until @jeromecoutant is back, or continue when he gets back?

mprse commented:
Please note that this PR contains only minor changes for rtc test.

cmonr commented:
@0xc0170 I'm thinking if @ARMmbed/team-st-mcd isn't able to respond by Wed, this PR should continue to move forward.

adbridge commented:
/morph build

adbridge commented:
@mprse  Couple of build errors:

Here is one of them
```
Compile [ 93.2%]: lp_ticker.c
        [DEBUG] Compile: /usr/local/ARM_Compiler_5.06u3/bin/armcc -c --gnu -Otime --split_sections --apcs=interwork --brief_diagnostics --restrict --multibyte_chars -O3 --cpu=Cortex-M3 --md --no_depend_system_headers --c99 -D__ASSERT_MSG -DMBED_STACK_STATS_ENABLED=1 -D__MBED__=1 -DTARGET_LIKE_MBED -DDEVICE_PORTINOUT=1 -D__MBED_CMSIS_RTOS_CM -DMBED_TRAP_ERRORS_ENABLED=1 -DMBED_BUILD_TIMESTAMP=1520332638.34 -D__CMSIS_RTOS -DTARGET_ARM_IOTSS_Target -DTARGET_LIKE_CORTEX_M3 -DTARGET_CORTEX_M -DARM_MATH_CM3 -DTARGET_UVISOR_UNSUPPORTED -DTARGET_M3 -DMBED_HEAP_STATS_ENABLED=1 -DDEVICE_INTERRUPTIN=1 -DTARGET_CORTEX -DDEVICE_I2C=1 -DDEVICE_PORTOUT=1 -D__CORTEX_M3 -DDEVICE_PORTIN=1 -DTARGET_RELEASE -DTARGET_CM3DS_MPS2 -DTARGET_ARM_SSG -DTOOLCHAIN_object -DCMSDK_CM3DS -DDEVICE_SPI=1 -DDEVICE_ETHERNET=1 -DDEVICE_ANALOGIN=1 -DDEVICE_SERIAL=1 -DTOOLCHAIN_ARM -DTOOLCHAIN_ARM_STD -DTARGET_ARM_CM3DS_MPS2 --via /builds/ws/mbed-os-build-matrix/1358/ARM_CM3DS_MPS2_ARM/sources/mbed-os/BUILD/tests/ARM_CM3DS_MPS2/ARM/.includes_15cabf29c3ab5c2c63b2efeab3031c61.txt --preinclude=/builds/ws/mbed-os-build-matrix/1358/ARM_CM3DS_MPS2_ARM/sources/mbed-os/BUILD/tests/ARM_CM3DS_MPS2/ARM/mbed_config.h --depend BUILD/tests/ARM_CM3DS_MPS2/ARM/targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/lp_ticker.d -o BUILD/tests/ARM_CM3DS_MPS2/ARM/targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/lp_ticker.o /builds/ws/mbed-os-build-matrix/1358/ARM_CM3DS_MPS2_ARM/sources/mbed-os/targets/TARGET_ARM_SSG/TARGET_CM3DS_MPS2/lp_ticker.c
        [Error] lp_ticker.c@58,0:  #20: identifier "lp_ticker_irq_handler" is undefined
        [Error] lp_ticker.c@90,0:  #20: identifier "timestamp_t" is undefined
```

mprse commented:
Fixed build errors on ARM_CM3DS_MPS2 and MIMXRT1050_EVK. @adbridge can we try again?

adbridge commented:
/morph build

mprse commented:
> Test : FAILURE
> 
> Build number : 1139
> Test logs :http://mbed-os-logs.s3-website-us-west-1.amazonaws.com/?prefix=logs/6161/1139

**UBLOX_EVK_ODIN_W2**: RTC tests failed on this platform. This boards inherits from STM32_FAMILY, but it looks like its RTC drivers are not consistent with the new standards. Since I do not have this board and so I cannot investigate the problem I suggest to disable it for now. We need two platforms enabled as soon as possible and according to the plan we will have NUCLEO boards (this PR) and K64F(PR https://github.com/ARMmbed/mbed-os/pull/6243) enabled - which is enough.

**NUCLEO_F401RE, NUCLEO_F429ZI, NUCLEO_F746ZG**: `tests-mbed_hal-rtc_reset` test fails. It looks like there is some kind of communication issue (@jeromecoutant looks like this is the same issue as you have encountered while testing RTC driver fix). I was able to reproduce the problem and get the same output while running `tests-mbed_hal-rtc_reset` test on board which has been just powered.  In such case RTC initialization takes more time than in futher `rtc_init` calls. This has impact on green-tea communication. Commands send by host immediately after init command (write, read) are not handled on the device side and no response to the host was provided.
To fix this problem test communication flow needs to be modified. Device must send ack to the host after RTC init is done. Host then sends further RTC commands when ack from the device has been received.

mprse commented:
Described updates have been implemented. @adbridge  Can we try again?

0xc0170 commented:
/morph build

mprse commented:
> Test : FAILURE
> 
> Build number : 1159
> Test logs :http://mbed-os-logs.s3-website-us-west-1.amazonaws.com/?prefix=logs/6161/1159

NUCLEO_F401RE(IAR), NUCLEO_F429ZI (GCC_ARM, IAR), NUCLEO_F746ZG(ALL): `tests-mbed_hal-rtc_reset` still fails. Looks like there are some communication issues on CI since such failures can not be reproduced locally. What we can try is to send `ack` from the device to host after each executed command and send next command from host after the `ack` is received. Will provide fix tomorrow.

cmonr commented:
@studavekar ^^^

studavekar commented:
> @studavekar ^^^
@cmonr ? :)

mprse commented:
@studavekar @cmonr Thanks for support.
Since `ack` helped to move the test forward on CI after init, I added `ack` after each command on the device side is executed/finished. So the next command is send from host after `ack` signal is received.

@0xc0170 Can we try again?

studavekar commented:
/morph build

0xc0170 commented:
@mprse Is this ready? The latest tests passed.

mprse commented:
> @mprse Is this ready? The latest tests passed.

Yes, its ready!
